types/src/validator_verifier.rs

```diff
@@ -159,75 +184,69 @@ impl<PublicKey: VerifyingKey> ValidatorVerifier<PublicKey> {
         Ok(())
     }
 
-    /// Ensure there are at least quorum_size and not more than maximum expected signatures.
-    fn check_num_of_signatures<T>(
+    /// Ensure there is at least quorum_voting_power in the provided signatures and there
+    /// are only known authors. According to the threshold verification policy,
+    /// invalid public keys are not allowed.
+    pub fn check_voting_power_and_signatures<'a>(
         &self,
-        aggregated_signature: &HashMap<AccountAddress, T>,
-    ) -> std::result::Result<(), VerifyError>
-    where
-        T: Into<PublicKey::SignatureMaterial> + Clone,
-    {
-        let num_of_signatures = aggregated_signature.len();
-        if num_of_signatures < self.quorum_size {
-            return Err(VerifyError::TooFewSignatures {
-                num_of_signatures,
-                quorum_size: self.quorum_size,
-            });
+        aggregated_signature: impl Iterator<Item = &'a AccountAddress>,
+    ) -> std::result::Result<(), VerifyError> {
+        let mut aggregated_voting_power = 0;
+        for account_address in aggregated_signature {
+            match self.get_voting_power(&account_address) {
+                Some(voting_power) => aggregated_voting_power += voting_power,
+                None => {
+                    return Err(VerifyError::UnknownAuthor);
+                }
+            }
         }
```

Comment:
You could possibly do instead:
```
let aggregated_voting_power = aggregated_signature.try_fold(0, |(acc, account_address)| {
    self.get_voting_power(&account_address).map(|vp| acc + vp)
}).unwrap_or_else(|| return Err(VerifyError::UnknownAuthor));
```
(note the case there `aggregated_signature.size() == 0` would then error instead of continue, though)

Reply:
I think this approach would require a bit more work (tried it out, but aggregated_voting_power is u64 and return is a Result), prefer the previous approach unless you feel strongly.

Reply:
Precisely this works:
```
        let aggregated_voting_power = aggregated_signature
            .into_iter()
            .try_fold(0, |acc, account_address| {
                self.get_voting_power(&account_address).map(|vp| acc + vp)
            })
            .ok_or_else(|| VerifyError::UnknownAuthor)?;
```

Reply:
Precisely this works:
```
        let aggregated_voting_power = aggregated_signature
            .into_iter()
            .try_fold(0, |acc, account_address| {
                self.get_voting_power(&account_address).map(|vp| acc + vp)
            })
            .ok_or_else(|| VerifyError::UnknownAuthor)?;
```

Reply:
Clippy isn't happy.  I'm inclined to leave this as this is a minor style difference.

   --> types/src/validator_verifier.rs:196:39
    |
196 |           let aggregated_voting_power = authors
    |  _______________________________________^
197 | |             .into_iter()
    | |________________________^ help: consider removing `.into_iter()`: `authors`